HHH-12944 - MultiIdentifierLoadAccess ignores the 2nd level cache

vladmihalcea opened the pull request:
https://hibernate.atlassian.net/browse/HHH-12944

vladmihalcea commented:
@sebersole if you have time, can you take a look at this PR? Thanks.

vladmihalcea commented:
Sure thing. I'll add the 2nd level cache enabling functionality, and maybe one to enable both with a single call. I'll also update the documentation to make it clear the outcome of each option.

sebersole commented:
> Sure thing. I'll add the 2nd level cache enabling functionality, and maybe one to enable both with a single call. I'll also update the documentation to make it clear the outcome of each option.

Sounds reasonable

vladmihalcea commented:
@sebersole I applied your suggestions and added more tests. Let me know what you think.

sebersole commented on hibernate-core/src/main/java/org/hibernate/MultiIdentifierLoadAccess.java:
```diff
@@ -69,6 +69,31 @@
+	 *
+	 * @return {@code this}, for method chaining
+	 */
+	MultiIdentifierLoadAccess<T> enableSessionFactoryCheck(boolean enabled);
```
How is this different than `MultiIdentifierLoadAccess<T> with(CacheMode cacheMode)`, which we already have?  This should just go away imo

sebersole commented on hibernate-core/src/main/java/org/hibernate/MultiIdentifierLoadAccess.java:
```diff
@@ -69,6 +69,31 @@
+	 *
+	 * @return {@code this}, for method chaining
+	 */
+	default MultiIdentifierLoadAccess<T> enableCacheCheck(boolean enabled) {
```
Again, I am not sure of this combined option.  At the *very least* the Javadoc (and documentation) should indicate that this is really just a limited "convenience" form of coordinating both `#with(CacheMode)` and `#enableSessionCheck`

vladmihalcea commented:
@sebersole I removed the `enableSessionFactoryCheck` and used `with(CacheMode cacheMode)`  instead. I also removed the `enableCacheCheck` method as it was not clear what `CacheMode` to use in this case.

I also updated the documentation to include some examples, including loading from the current Session or SessionFactory.

Let me know what you think.

sebersole commented on documentation/src/main/asciidoc/userguide/chapters/pc/PersistenceContext.adoc:
```diff
@@ -174,6 +174,111 @@ include::{sourcedir}/PersistenceContextTest.java[tags=pc-find-optional-by-id-nat
+In other words, consumers of the returned ordered List would need to be able to handle null elements.
+`enableSessionCheck(boolean enabled)`::
+This setting, which is disabled by default, tells Hibernate to check the first-level cache (a.k.a `Session` or Persistence Context) first and skip entities that are already loaded from being included
+in the multi-load query.
```
The thing people keep harping about ;)  Backward compatible behavior.  That was the original behavior.  

You are probably correct about that being the natural expectation.  But keep in mind also that "ordering" also implies other things that the original code just did not do.  For example, initially loading `[1,2,2,2,3]` would get you a unique list - `[Thing#1, Thing#2, Thing#3]`.  But if the incoming list order is to be maintained we have to know to add the duplicates and return `[Thing#1, Thing#2, Thing#2, Thing#2, Thing#3]`

Also, the ordering comes with a performance cost.  

So given all that, I just did not see enough to change the default from the backwards-compatible behavior once it was made configurable.

gsmet replied:
I was talking about checking the persistence context cache, not about the ordering.

Or does checking the persistence context has consequence on the result?

sebersole replied:
Oh, I mis-read.  Here though too, the original behavior was as described.  And here too there is a performance trade-off.  It made sense to add both options, so we did, but again decided to leave the default behavior as the backwards compatible one.

It does not affect whether an entity is included or not - you are correct about that.  It just affects how we go about getting each of them loaded.

vladmihalcea replied:
Related to wording, I changed the description above to:

> This setting, which is disabled by default, tells Hibernate to check the first-level cache (a.k.a `Session` or Persistence Context) first and, if the entity is found and already managed by the Hibernate `Session`, the cached entity will be added to the result set, therefore skipping it from being fetched via the multi-load query.

sebersole commented on documentation/src/test/java/org/hibernate/userguide/pc/MultiLoadIdTest.java:
```diff
@@ -0,0 +1,167 @@
+	}
+
+	@Test
+	public void testSessionFactoryCheck() {
```
I'll go out on a limb and answer for @vladmihalcea , but probably for the same reason that checking-session is accepted when really its checking the "first level cache" (PersistenceContext) instead.

gsmet replied:
Hum. Yeah. Never heard of a session factory cache though. Whereas a session cache kinda makes sense.

I think 2nd level cache would be clearer for the users.

sebersole replied:
Fair enough.  Not arguing either way.

Just pointing out that its not really as "weird" as you made it out to be.  

Generally speaking, I think the users that understand these terms at all are going to understand either variation.

vladmihalcea commented on hibernate-core/src/main/java/org/hibernate/internal/SessionImpl.java:
```diff
@@ -2969,6 +2969,11 @@ public boolean isSessionCheckingEnabled() {
 		}
 
+		@Override
+		public boolean isSessionFactoryCheckingEnabled() {
```
Just like the 1st-level cache is linked to the Session, the 2nd-level cache is linked to the SessionFactory. The name of the method was chosen to resemble the `isSessionCheckingEnabled` method we already have.

However, since we already have methods like `loadFromSessionCache` and `loadFromSecondLevelCache`, it's probably better to use `SecondLevelCache` in this context.

vladmihalcea commented on hibernate-core/src/main/java/org/hibernate/persister/entity/MultiLoadOptions.java:
```diff
@@ -14,11 +14,47 @@
 	boolean isReturnOfDeletedEntitiesEnabled();
+
+	/**
+	 * Should the entities be return in the same order as their associated entity identifiers were provided.
```
Fixed.

vladmihalcea commented on documentation/src/main/asciidoc/userguide/chapters/pc/PersistenceContext.adoc:
```diff
@@ -174,6 +174,111 @@ include::{sourcedir}/PersistenceContextTest.java[tags=pc-find-optional-by-id-nat
+`with(RootGraph<T> graph)`::
+The `RootGraph` is a Hibernate extension to the JPA `EntityGraph` contract,
+and this method allows you to pass a specific `RootGraph` to the multi-load query
+so that it can fetch addition relationships of the current loading entity.
```
Fixed
